packages/protocol/contracts/governance/Governance.sol

```diff
@@ -649,7 +672,71 @@ contract Governance is IGovernance, Ownable, Initializable, UsingLockedGold, Ree
   }
 
   /**
-   * @notice Withdraws refunded Celo Gold commitments.
+   * @notice Whitelists the hash of a hotfix.
+   * @param txHash The abi encoded keccak256 hash of the hotfix transaction to be whitelisted.
+   * @param validatorIndex The index in the validator set of the whitelisting validator.
+   */
+  function whitelistHotfix(bytes32 txHash, uint256 validatorIndex) external nonReentrant {
+    if (msg.sender == approver) {
+      require(!hotfixes[txHash].approved);
+      hotfixes[txHash].approved = true;
+    } else {
+      IValidators validators = IValidators(registry.getAddressForOrDie(VALIDATORS_REGISTRY_ID));
+      require(msg.sender == validators.validatorAddressFromCurrentSet(validatorIndex));
+      require(!hotfixes[txHash].whitelisted[msg.sender]);
+      hotfixes[txHash].whitelisted[msg.sender] = true;
+    }
+
+    emit HotfixWhitelisted(txHash, msg.sender);
+  }
+
+  /**
+   * @notice Gives hotfix a prepared epoch for execution.
+   * @param txHash The hash of the hotfix to be prepared.
+   */
+  function prepareHotfix(bytes32 txHash) external {
+    IValidators validators = IValidators(registry.getAddressForOrDie(VALIDATORS_REGISTRY_ID));
+    uint256 epoch = validators.getEpochNumber();
+    require(hotfixes[txHash].preparedEpoch < epoch, "hotfix already prepared for this epoch");
+    require(isHotfixPassing(txHash), "hotfix not whitelisted by 2f+1 validators");
+    hotfixes[txHash].preparedEpoch = epoch;
+    emit HotfixPrepared(txHash, epoch);
+  }
+
+  /**
+   * @notice Executes a whitelisted proposal.
+   * @param values The values of Celo Gold to be sent in the proposed transactions.
+   * @param destinations The destination addresses of the proposed transactions.
+   * @param data The concatenated data to be included in the proposed transactions.
+   * @param dataLengths The lengths of each transaction's data.
+   * @dev Reverts if hotfix is already completed, not approved, or not prepared for current epoch.
+   */
+  function executeHotfix(
+    uint256[] calldata values,
+    address[] calldata destinations,
+    bytes calldata data,
+    uint256[] calldata dataLengths
+  )
+    external
+  {
+    bytes32 txHash = keccak256(abi.encode(values, destinations, data, dataLengths));
```

Comment:
Consider defining a `hash` function on `Proposals.Proposal`

Reply:
IMO the `txHash` should be easily computable on the javascript side from the same inputs that are going to be passed to the `executeHotfix` function. If we make a `hash` function on `Proposals.Proposal`, it will have to do significant unpacking of the `Proposal` struct to compute the same encoding, or change the encoding to a harder to compute/verify pattern.